src/herder/TxSetFrame.h

```diff
@@ -71,65 +106,137 @@ class TxSetFrame : public NonMovableOrCopyable
 #endif
     );
 
+    // Creates a valid empty TxSetFrame pointing at provided `lclHeader`.
+    // The output TxSetFrame is not resolved as `lclHeader` doesn't necessarily
+    // match the real LCL.
+    static TxSetFrameConstPtr
+    makeEmpty(LedgerHeaderHistoryEntry const& lclHeader);
+
+    // `makeFromWire` methods create a TxSetFrame from the XDR messages.
+    // These methods don't perform any validation on the XDR. They have to
+    // be `resolve`d first, then validated with `checkValid`.
+    static TxSetFrameConstPtr makeFromWire(TransactionSet const& xdrTxSet);
+    static TxSetFrameConstPtr
+    makeFromWire(GeneralizedTransactionSet const& xdrTxSet);
+
+    static TxSetFrameConstPtr
+    makeFromStoredTxSet(StoredTransactionSet const& storedSet);
+
     // Creates a legacy (non-generalized) TxSetFrame from the transactions that
     // are trusted to be valid. Validation and filtering are not performed.
     // This should be *only* used for building the legacy TxSetFrames from
     // historical transactions.
     static TxSetFrameConstPtr
     makeFromHistoryTransactions(Hash const& previousLedgerHash,
-                                Transactions const& txs);
+                                TxSetFrame::Transactions const& txs);
 
-    // Creates a valid empty TxSetFrame.
-    static TxSetFrameConstPtr
-    makeEmpty(LedgerHeaderHistoryEntry const& lclHeader);
+    void toXDR(TransactionSet& set) const;
+    void toXDR(GeneralizedTransactionSet& generalizedTxSet) const;
+    void storeXDR(StoredTransactionSet& txSet) const;
 
-    // Creates a TxSetFrame from the XDR message.
-    // As the message is not trusted, it has to be validated via `checkValid`.
-    static TxSetFrameConstPtr makeFromWire(Application& app,
-                                           TransactionSet const& xdrTxSet);
-    static TxSetFrameConstPtr
-    makeFromWire(Application& app, GeneralizedTransactionSet const& xdrTxSet);
+    ~TxSetFrame() = default;
 
-    // Creates a TxSetFrame from StoredTransactionSet (internally persisted tx
-    // set format).
-    static TxSetFrameConstPtr
-    makeFromStoredTxSet(StoredTransactionSet const& storedSet,
-                        Application& app);
+    // Resolves this transaction set using the current ledger set.
+    // The resolution includes any logic dependent on the protocol version
+    // and network configuration state, i.e. any logic that depends on
+    // the current ledger state. That's why this may *only* be called
+    // when LCL hash matches the `previousLedgerHash` of this TxSetFrame.
+    //
+    // The returned value has the same lifetime as this `TxSetFrame` and
+    // is cached, so it's safe and cheap to call this multiple times.
+    ResolvedTxSetFrame const* resolve(Application& app,
```

Comment:
I'm not intending to share the ownership of the resolved tx frame. Const pointer is a typical way of representing an optional reference to an object. It's only marginally more dangerous than returning a const reference (that we do a lot), and the additional 'danger' comes from the fact that it's optional, so using any type of pointer/optional etc. would come with the same risk - it's responsibility of the caller to make sure it's non-nullptr.

Reply:
I explicitly do not want to share the ownership because only tx set frame owns its resolved part. Hence its a unique ptr.
Const pointer conveys that in the same way as const reference and is subject to the same issues (because it's C++ and references aren't that different from pointers). Again, if we have issue with returning const references, we should probably rewrite a good chunk of the code base. I don't really think that's an issue though, as long as we set the right expectations on the referred object's lifetime. Returning a weak ptr here forces conceptually incorrect ownership (because, again, it's not intended to be shared).
If you find pointers really misleading/offensive, we could use `optional<reference_wrapper<ResolvedTxSetFrame const>>` which would convey exactly the same thing, but in a pointer-less way (albeit more verbose). WDYT?